From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 8 Feb 2025 22:11:16 +0100
Subject: [PATCH] Remove stream in villagers

TradeWithVillager#figureOutWhatIAmWillingToTrade
In the test, this can give ~40% improvement (~20ms -> ~12ms),
under 2048 villagers situation.
And ~100% improvement (~36ms -> ~0ms), under 512 villagers situation.

Villager#countFoodPointsInInventory
In the test, this can give ~82.14% improvement (~1456ms -> ~260ms),
under 2048 villagers situation.
And ~93.92% improvement (~1382ms -> ~84ms), under 512 villagers situation.

Ported from Leaf: https://github.com/Winds-Studio/Leaf

diff --git a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
index 736ad21b73a19113c6f0f3bce6aa1cc0a41485a8..1a54bace55744032dae8d6ac9977de0680146c98 100644
--- a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
+++ b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
@@ -73,7 +73,15 @@ public class TradeWithVillager extends Behavior<Villager> {
     private static Set<Item> figureOutWhatIAmWillingToTrade(final Villager myBody, final Villager target) {
         ImmutableSet<Item> targetItems = target.getVillagerData().profession().value().requestedItems();
         ImmutableSet<Item> selfItems = myBody.getVillagerData().profession().value().requestedItems();
-        return targetItems.stream().filter(entry -> !selfItems.contains(entry)).collect(Collectors.toSet());
+        // Shulker start - Remove stream in villagers
+        Set<Item> ret = new java.util.HashSet<>();
+        for (Item item : targetItems) {
+            if (!selfItems.contains(item)) {
+                ret.add(item);
+            }
+        }
+        return ret;
+        // Shulker end - Remove stream in villagers
     }
 
     private static void throwHalfStack(final Villager villager, final Set<Item> items, final LivingEntity target) {
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index 0a02aef2ff159a1d94870bf3cede0ed29ae192af..74cadd3a84fc489f6a50c62ed13fadbd64ae220b 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -844,7 +844,17 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Shulker start - Remove stream in villagers
+        int sum = 0;
+
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            Item item = entry.getKey();
+            int points = entry.getValue();
+            sum += inventory.countItem(item) * points;
+        }
+
+        return sum;
+        // Shulker end - Remove stream in villagers
     }
 
     public boolean hasFarmSeeds() {
